### PR TITLE
[WIP] Introduce optionalfields linter

### PR DESCRIPTION
**Pull request by JoelSpeed**
This adds a new linter for `optionalfields` that checks optional fields for things like `omitempty` and whether zero values are valid or not.

The linter, depending on configuration can:
* Make sure you have omitempty on optional fields
* Make sure you have a pointer on optional fields
* Where you don't require pointers:
  * Request markers for length/limits to be able to determine if the zero value is valid, and therefore whether the field should actually be a pointer or not
  * Identify structs that do or do not have required fields, and therefore should/should not have pointers
* Where omitempty is not desired, it can be ignored, and the rules change slightly to make sure the zero value marshalled is an allowable value

### CODE REVIEW

**Comment by JoelSpeed** [comment]
I think this is now ready for some initial reviews and testing.

One thing I've discovered, which we will need to look into in more depth, is that currently, when types are imported, or are even within a different file within the same package, we aren't able to pick up and lint references of fields to those types.

So there are going to be things we miss, and this is going to affect the wider project right now until we come up with a solution. I think the solution will be in the shared inspector code though, and I plan to investigate more tomorrow.

**Comment by sivchari** [comment]
I'll review it tomorrow (It's 8:00 PM JST 😄 )

File: pkg/analysis/optionalfields/util.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+	return ptr.Deref(minimum, -1) == 0 ||
+		ptr.Deref(maximum, -1) == 0 ||
+		ptr.Deref(minimum, 0) < 0 && ptr.Deref(maximum, 0) > 0
+}
+
+func floatRangeIncludesZero(minimum, maximum *float64) bool {
+	return ptr.Deref(minimum, -1) == 0 ||
+		ptr.Deref(maximum, -1) == 0 ||
+		ptr.Deref(minimum, 0) < 0 && ptr.Deref(maximum, 0) > 0
+}
```
Comment (everettraven) [inline pkg/analysis/optionalfields/util.go:273]: Nit: This _could_ use generics since this is a pretty straightforward "does range of min/max contain 0"

File: pkg/config/linters_config.go [hunk 2]
```diff
@@ -139,6 +142,80 @@ type NoMapsConfig struct {
...
+	// preference determines whether the linter should prefer pointers for all optional fields,
+	// or only for optional fields where validation or serialization requires a pointer.
+	// Valid values are "Always" and "WhenRequired".
+	// When set to "Always", the linter will prefer pointers for all optional fields.
+	// When set to "WhenRequired", the linter will prefer pointers for optional fields where validation or serialization requires a pointer.
+	// The "WhenRequired" option requires bounds on strings and numerical values to be able to acurately determine the correct pointer vs non-pointer decision.
+	// When otherwise not specified, the default value is "Always".
+	Preference OptionalFieldsPointerPreference `json:"preference"`
+	// policy is the policy for the pointer preferences for optional fields.
+	// Valid values are "SuggestFix" and "Warn".
```
Comment (everettraven) [inline pkg/config/linters_config.go:169]: Ignore no longer a valid value?

File: README.md [hunk 3]
```diff
@@ -230,6 +230,28 @@ lintersConfig:
...
+### Configuration
+
+```yaml
+lintersConfig:
+  optionalFields:
+    pointers:
+      preference: Always | WhenRequired # Whether to always require pointers, or only when required. Defaults to `Always`.
+      policy: SuggestFix | Warn | Ignore # The policy for pointers in optional fields. Defaults to `SuggestFix`.
+    omitempty:
+        policy: SuggestFix | Warn | Ignore # The policy for omitempty in optional fields. Defaults to `SuggestFix`.
```
Comment (everettraven) [inline README.md:252]: Looks like this may be outdated based on what is actually implemented? I.e `pointers.policy` doesn't allow `Ignore`  and `omitempty.policy` doesn't allow `Warn`

**Review by everettraven — COMMENTED** [review]
Feel free to ignore any nits if you feel they aren't worth tackling

File: README.md [hunk 4]
```diff
@@ -261,6 +261,28 @@ lintersConfig:
...
+In this case, the `omitempty` policy can be set to `Ignore`, and the linter will ensure that the zero value of the object is an acceptable value for the field.
+
+### Configuration
+
+```yaml
+lintersConfig:
+  optionalFields:
+    pointers:
+      preference: Always | WhenRequired # Whether to always require pointers, or only when required. Defaults to `Always`.
+      policy: SuggestFix | Warn # The policy for pointers in optional fields. Defaults to `SuggestFix`.
```
Comment (everettraven) [inline README.md:281]: Should we include any additional information in this section about the policy options? I think the names make sense , but a short sentence about what each mode means might help to eliminate any ambiguity - especially if we introduce different policy-like fields in the future that deviate from this terminology (for whatever reason).

File: pkg/analysis/optionalfields/analyzer.go [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+	requiredMarker            = "required"
+	kubebuilderOptinalMarker  = "kubebuilder:validation:Optional"
+	kubebuilderRequiredMarker = "kubebuilder:validation:Required"
+
+	minItemsMarker      = "kubebuilder:validation:MinItems"
+	minLengthMarker     = "kubebuilder:validation:MinLength"
+	minPropertiesMarker = "kubebuilder:validation:MinProperties"
+
+	minimumMarker = "kubebuilder:validation:Minimum"
+	maximumMarker = "kubebuilder:validation:Maximum"
```
Comment (everettraven) [inline pkg/analysis/optionalfields/analyzer.go:47]: Use the recently merged markers package instead of defining new constants

File: pkg/analysis/optionalfields/analyzer.go [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+
+	fieldName := field.Names[0].Name
+
+	if !isFieldOptional(fieldMarkers) {
+		// The field is not marked optional, so we don't need to check it.
+		return
+	}
+
+	if field.Type == nil {
+		// The field has no type? We can't check if it's a pointer.
```
Comment (everettraven) [inline pkg/analysis/optionalfields/analyzer.go:115]: Out of curiosity, what could cause a scenario like this to happen? In this scenario could we still check for the omitempty and skip only the pointer check?

File: pkg/analysis/optionalfields/analyzer.go [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+
+	switch a.pointerPreference {
+	case config.OptionalFieldsPointerPreferenceAlways:
+		a.checkFieldPointersPreferenceAlways(pass, field, fieldName, isStarExpr)
+	case config.OptionalFieldsPointerPreferenceWhenRequired:
+		a.checkFieldPointersPreferenceWhenRequired(pass, field, fieldName, isStarExpr, underlyingType, markersAccess, jsonTags)
+	}
+}
+
+// checkFieldPointersPointerTypes checks for poitner types (eg Maps and Arrays)
```
Comment (everettraven) [inline pkg/analysis/optionalfields/analyzer.go:171]:
```suggestion
// checkFieldPointersPointerTypes checks for pointer types (eg Maps and Arrays)
```

**Review by everettraven — COMMENTED** [review]
A couple more minor comments and some tweaks based on changes that have recently gone into main, but other than that I think this looks good.

File: README.md (hunk 4 above)
Comment (JoelSpeed) [reply README.md:299]: I forgot to add the `Fixes` title, adding now

File: pkg/analysis/optionalfields/analyzer.go [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+		}
+
+		a.checkFieldPointersPreferenceWhenRequiredIdentObj(pass, field, fieldName, isStarExpr, decl, markersAccess, jsonTags)
+
+		return
+	}
+
+	switch ident.Name {
+	case "int", "int8", "int16", "int32", "int64", "uint", "uint8", "uint16", "uint32", "uint64":
+		a.checkFieldPointersPreferenceWhenRequiredInteger(pass, field, fieldName, isStarExpr, markersAccess, jsonTags)
```
[the remark this answers is not in the document]
Comment (everettraven) [reply pkg/analysis/optionalfields/analyzer.go:248]: You would need to create your own generic type like:
```go
type Number interface {
    int | int8 | int16 | int32 | int64 | uint | ...
}
```

I don't know if there already exists something in the std lib that would make this easier.

Not a huge deal though.

[the remark this answers is not in the document]
Comment (everettraven) [reply pkg/analysis/optionalfields/analyzer.go:237]: Totally fine. I'm not blocking this on that.

**Comment by JoelSpeed** [comment]
/hold in the logic for whether an optional struct can be added without omitempty or not, we need to take into account the zero marshalled struct. To do this we need to look at the fields within and determine if they would be marshalled as their zero values. If they are, and the zero value is not acceptable (through some sort of validation), then the parent struct cannot be pointer-less/omitempty-less

I will try and work up an update

File: pkg/analysis/optionalfields/testdata/src/c/a.go [hunk 9]
```diff
@@ -129,6 +129,11 @@ type A struct {
 	// +optional
 	StringWithMinLength0 string `json:"stringWithMinLength0,omitempty"` // want "field StringWithMinLength0 has a minimum length of 0. The empty string is a valid value and therefore the field should be a pointer"
 
+	// EnumString is a string field with an enum.
+	// +kubebuilder:validation:Enum=foo;bar;baz
+	// +optional
+	EnumString string `json:"enumString,omitempty"`
```
Comment (everettraven) [inline pkg/analysis/optionalfields/testdata/src/c/a.go:135]: Can we also add a test for the case of an enum allowing the empty value and having the omitempty tag? 

In that case, based on the code, it should signal that the field should be a pointer to the type.

**Review by everettraven — COMMENTED** [review]
Other than the one comment about adding a test for the additional enum case, I think this is good to go

**Comment by JoelSpeed** [comment]
/approve cancel

I'm in the process of rewriting this, it got way too complex and I think we can be more methodical about making decisions and be more useful by err'ing on the side of caution when recommending pointers

File: pkg/analysis/optionalfields/util.go [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+	var numberValue N
+	switch any(&numberValue).(type) {
+	case *int:
+		numberValue = N(int(value))
+	case *float64:
+		numberValue = N(value)
+	default:
+		// Panic here since this error is a programming error.
+		panic(fmt.Sprintf("unsupported number type: %T", numberValue))
+	}
```
Comment (everettraven) [inline pkg/analysis/optionalfields/util.go:403]: Is there a particular reason that doing a `return N(value)` is insufficient here?

Having a switch on types is usually an anti-pattern when using generics.

I played around with this and it seems like a `return N(value)` _should_ work - https://go.dev/play/p/U_03gG-Fcl8 (might be a bit naive, but I think this is representative of the scenario here)

File: pkg/analysis/optionalfields/analyzer.go [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+	kubebuilderRequiredMarker = markers.KubebuilderRequiredMarker
+
+	minItemsMarker      = markers.KubebuilderMinItemsMarker
+	minLengthMarker     = markers.KubebuilderMinLengthMarker
+	minPropertiesMarker = markers.KubebuilderMinPropertiesMarker
+
+	minimumMarker = markers.KubebuilderMinimumMarker
+	maximumMarker = markers.KubebuilderMaximumMarker
+
+	enumMarker = markers.KubebuilderEnumMarker
```
Comment (everettraven) [inline pkg/analysis/optionalfields/analyzer.go:49]: Any particular reason to make these local constants instead of just using the `markers` package constants?

**Comment by JoelSpeed** [comment]
/approve

This could be improved, as in the case where you import types from external sources, it errs on the side of caution (optional should be pointer and omitempty), but, I think that can be a follow up.
